src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs

```diff
@@ -0,0 +1,365 @@
+                {
+                    item.Json.Dispose();
+                }
+            }
+        }
+
+        public static bool IsOutputDocumentPrefix(string prefix)
+        {
+            var parts = prefix.Split("/", StringSplitOptions.RemoveEmptyEntries);
+
+            if (parts.Length < 2)
+                return false;
+
+            string numeric = parts.Last();
+
+            if (long.TryParse(numeric, out _) == false)
+                return false;
+
+            return true;
+        }
+
+        public static string GetOutputDocumentPrefix(string collectionName, long reduceOutputIndex)
+        {
+            return collectionName + "/" + reduceOutputIndex + "/";
+        }
+
+        private string GetOutputDocumentKey(string reduceKeyHash)
+        {
+            if (_reduceOutputIndex != null)
+                return GetOutputDocumentPrefix(_outputReduceToCollection, _reduceOutputIndex.Value) + reduceKeyHash;
+
+            return _outputReduceToCollection + "/" + reduceKeyHash;
+        }
+
+        private string GetOutputDocumentKeyForSameReduceKeyHash(string reduceKeyHash, int outputNumber)
+        {
+            return GetOutputDocumentKeyForSameReduceKeyHashPrefix(reduceKeyHash) + outputNumber;
+        }
+
+        private string GetOutputDocumentKeyForSameReduceKeyHashPrefix(string reduceKeyHash)
+        {
+            return GetOutputDocumentKey(reduceKeyHash) + MultipleOutputsForSameReduceKeyHashSeparator;
+        }
+
+        public void Dispose()
+        {
+            foreach (var r in _reduceDocuments)
+            {
+                foreach (var doc in r.Value)
+                {
+                    doc.Json.Dispose();
+                }
+            }
+        }
+
+        private class ReduceOutputReferencesSubCommand
+        {
+            private readonly MapReduceIndex _index;
+            private readonly string _outputReduceToCollection;
+            private readonly Transaction _indexWriteTransaction;
+            private readonly DocumentDatabase _database;
+            readonly Dictionary<string, HashSet<string>> _referencesOfReduceOutputs = new Dictionary<string, HashSet<string>>();
+            readonly Dictionary<string, HashSet<string>> _idsToDeleteByReferenceDocumentId = new Dictionary<string, HashSet<string>>();
+
+            public ReduceOutputReferencesSubCommand(MapReduceIndex index, string outputReduceToCollection, Transaction indexWriteTransaction)
+            {
+                _index = index;
+                _outputReduceToCollection = outputReduceToCollection;
+                _indexWriteTransaction = indexWriteTransaction;
+                _database = index.DocumentDatabase;
+            }
+
+            public void Add(string referenceDocumentId, string reduceOutputDocumentId)
+            {
+                if (_referencesOfReduceOutputs.TryGetValue(referenceDocumentId, out var values) == false)
+                {
+                    values = new HashSet<string>(1);
```

Comment:
case insensitive match?

Reply:
I used `StringComparer.OrdinalIgnoreCase` in all placed where I use HashSet or Dictionary

Reply:
But not here ,though?

Reply:
Here as well. Why not? Document IDs are case insensitive - am I missing something?

Reply:
`new HashSet<string>(1)` will do case sensitive comparisons

Reply:
Oh, I meant that I changed it locally to `new HashSet<string>(1, StringComparer.OrdinalIgnoreCase);`